File: utilities/src/main/java/xerus/util/helpers/ArraySet.java
package xerus.util.helpers;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Set;

/** Implementation of ArrayList which prohibits double Elements */
public class ArraySet<E> extends ArrayList<E> implements Set<E> {
	
	public ArraySet() {
		super();
	}
	
	public ArraySet(Collection<E> list) {
		super(list);
	}
	
	@Override
	public boolean add(E e) {
		return !contains(e) && super.add(e);
	}
	
	@Override
	public void add(int index, E e) {
		if (!contains(e)) {
			super.add(index, e);
		}
	}
	
	@Override
	public boolean addAll(Collection<? extends E> c) {
		return addAll(size(), c);
	}
	
	@Override
	public boolean addAll(int index, Collection<? extends E> c) {
		Collection<E> copy = new ArrayList<E>(c);
		copy.removeAll(this);
		return super.addAll(index, copy);
	}
	
}

File: kotlin/src/xerus/ktutil/helpers/ArraySet.kt
package xerus.ktutil.helpers

import java.util.*

/** Implementation of ArrayList which prohibits double Elements  */
class ArraySet<E>() : ArrayList<E>(), Set<E> {
	
	constructor(list: Collection<E>) : this() {
		addAll(list)
	}
	
	override fun add(element: E): Boolean {
		return if (contains(element)) false else super.add(element)
	}
	
	override fun add(index: Int, element: E) {
		if (!contains(element))
			super.add(index, element)
	}
	
	override fun addAll(elements: Collection<E>) = addAll(size, elements)
	
	override fun addAll(index: Int, elements: Collection<E>): Boolean {
		val copy = ArrayList(elements)
		copy.removeAll(this)
		return super.addAll(index, copy.distinct())
	}
	
	override fun spliterator() = super<ArrayList>.spliterator()
	
}